feat(ux): keep app running when main window closed

mrfelton opened the pull request:
The standard approach on Darwin is to keen applications running when the window is closed. Clicking the apps icon again in the doc should bring back the application window.

This changeset refactors things to properly handle this. It will keep Zap and lnd running in the background when the main application window is closed. The gRPC connection is closed out when the window is closed and a new gRPC connection is established again once it has been reopened. Quitting the app will stop both the app and lnd.

I think there are further improvements that can be made to the user experience here as part of https://github.com/LN-Zap/zap-desktop/issues/520

Empact commented on app/lnd/lib/lightning.js:
```diff
@@ -21,37 +26,111 @@ process.env.GRPC_SSL_CIPHER_SUITES =
+   */
+  subscribeToTransactions() {
+    this.subscriptions.transactions = subscribeToTransactions(this.mainWindow, this.lnd, mainLog)
+  }
```
How about inlining these into `subscribe` to minimize the interface. `subscriptions` could then be an array produced by a map.

mrfelton replied:
I'm not quite sure what you mean about making subscriptions an array produced by a map, but I have refactored this to reduce the interface so that there is only the one `subscribe` method

JimmyMow commented on app/main.dev.js:
```diff
@@ -132,4 +126,16 @@ app.on('ready', () => {
+   * Add application event listener:
+   *  - reopen window when dock icon is clicked.
+   */
+  app.on('activate', () => {
```
nice!

JimmyMow commented:
utACK 54b7cb1

@mrfelton can you rebase this so I can test more accurately?

mrfelton commented:
Hmm. It seems this isn't working as well as it did before for some reason. Leave this one with me for a while @JimmyMow

JimmyMow commented:
👍

mrfelton commented:
@JimmyMow I've updated this PR to improve the way that this works somewhat.

I have implemented [electron-redux](https://www.npmjs.com/package/electron-redux) to enable us to share application state between windows. This gives us a single state that is initialised from the main process and syncronised with every browser window subproces. This allows us to open a new browser window with the previous application state in place for a more seamless user experience on the mac.

Now, if you close a window everything keeps running in the background (except the gRPC connection which we terminate). When you click on the app icon in the doc it will open a new window, reconnect gRPC, load the most recent data, and return the user to their already logged in wallet view.

JimmyMow commented on app/lnd/subscribe/channelgraph.js:
```diff
@@ -1,8 +1,14 @@
 
   call.on('data', channelGraphData => mainWindow.send('channelGraphData', { channelGraphData }))
+  call.on('end', () => log.info('end'))
+  call.on('error', error => (error.code === status.CANCELLED ? null : log.error(error)))
```
this can be simplified a bit because you're not utilizing the front half of the conditional:

`call.on('error', error => (error.code !== status.CANCELLED && log.error(error)))`

JimmyMow commented on app/lnd/subscribe/invoices.js:
```diff
@@ -1,8 +1,15 @@
   call.on('end', () => log.info('end'))
-  call.on('error', error => log.error(error))
-  call.on('status', status => log.info('status:', status))
+  call.on('error', error => (error.code === status.CANCELLED ? null : log.error(error)))
```
Same as the other comment, this can be simplified

JimmyMow commented on app/lnd/subscribe/transactions.js:
```diff
@@ -1,10 +1,15 @@
   })
   call.on('end', () => log.info('end'))
-  call.on('error', error => log.error('error: ', error))
+  call.on('error', error => (error.code === status.CANCELLED ? null : log.error(error)))
```
Can be simplified
